_drafts/best-of-binutils.md

```diff
@@ -0,0 +1,1011 @@
+  [15] .debug_str        PROGBITS        00000000 0382c1 00dedb 01  MS  0   0  1
+  [18] .debug_frame      PROGBITS        00000000 046248 000d8c 00      0   0  4
+
+# Remove debug info
+$ arm-none-eabi-strip -g nrf52_example/build/nrf52.elf
+
+# Confirm ELF no longer has debug info
+$ arm-none-eabi-readelf -S nrf52_example/build/nrf52.elf | grep debug
+
+```
+
+### objcopy
+
+`objcopy` is most commonly used as the last stage of a build to generate the `.bin` or `.hex` file which
+can be flashed onto a device. Unlike an operating system where there is a "dynamic loader" that
+resolves addresses and loads libaries at runtime, on embedded devices all of the memory locations
+are typically resolved at compilation time.
+
+```bash
+$ arm-none-eabi-objcopy firmware.elf firmware.bin -O binary
+```
+
+> NOTE: For an example of leveraging using objcopy to build a bootloader for an embedded device,
+> check out [this post]({% post_url 2019-08-13-how-to-write-a-bootloader-from-scratch %}#putting-it-all-together).
+
+#### Converting binaries into an object file
+
+Sometimes for embedded development you may want to include binary artifacts into a firmware as a
+byte array. For example, sometimes a firmware will include binaries that need to be loaded onto peripherals (i.e a GPS,
+or accelerometer peripheral image). One neat thing you can do with objcopy is convert any binary blob into an object file that can
+then be linked in your image.
+
+For example, if you want to include `my_binary.bin` in an image, you could create a `.o` as
+follows:
+
+```bash
+$ arm-none-eabi-objcopy -I binary -O elf32-littlearm
+   --add-section my_binary_section=my_binary.bin my_binary.bin my_binary.o
+```
+
+`objcopy` will emit symbols in the object that then make it easy to reference the blob from your
+C Code:
+
+```bash
+arm-none-eabi-nm -S -l my_binary.o
+00000031 D _binary_my_binary_bin_end
+00000031 A _binary_my_binary_bin_size
+00000000 D _binary_my_binary_bin_start
+```
+
+### addr2line
+
+`addr2line` converts addresses back to function and line numbers.
+
+#### Useful Arguments
+
+- `-a` To show the address which was decoded. This can be useful when decoding multiple addresses
+  at once or using `addr2line` in a batch processing mode.
+- `-f` To display the function name the address is in
+- `-e <ELF>` To specify the excutable to decode. (Otherwise, `addr2line` checks for `a.out` in the
+  working directory and uses that.)
+
+#### Example Usage
+
+You can run it on a single address:
+
+```bash
+$ arm-none-eabi-addr2line -a -f -e build/zephyr/zephyr.elf 0x08000946
+  0x08000946
+_printk_dec_ulong
+zephyr/lib/os/printk.c:465
+
+```
+
+or pipe a bunch of addresses in from stdin and get them all decoded at once
+
+```bash
+$ echo -e "0x8000946\n0x8000336" |
+   arm-none-eabi-addr2line -a -f -e build/zephyr/zephyr.elf
+0x08000946
+_printk_dec_ulong
+zephyr/lib/os/printk.c:465
+0x08000336
+__udivmoddi4
+```
+
```

Comment:
Plug for Memfault maybe: If you find yourself writing scripts to parse addresses from a crash output, you might like to try Memfault.